components/imageproc/src/lib.rs

```diff
@@ -459,9 +529,89 @@ impl Processor {
-                op.perform(&target)
-                    .map_err(|e| Error::chain(format!("Failed to process image: {}", op.source), e))
+                op.perform(&target).map_err(|e| {
+                    Error::chain(format!("Failed to process image: {}", op.input_path.display()), e)
+                })
             })
             .collect::<Result<()>>()
     }
 }
+
+#[derive(Debug, Serialize, Eq, PartialEq)]
+pub struct ImageMetaResponse {
+    pub width: u32,
+    pub height: u32,
+    pub format: Option<&'static str>,
+}
+
+impl ImageMetaResponse {
+    pub fn new_svg(width: u32, height: u32) -> Self {
+        Self { width, height, format: Some("svg") }
+    }
+}
+
+impl From<ImageMeta> for ImageMetaResponse {
+    fn from(im: ImageMeta) -> Self {
+        Self {
+            width: im.size.0,
+            height: im.size.1,
+            format: im.format.and_then(|f| f.extensions_str().get(0)).map(|&f| f),
+        }
+    }
+}
+
+impl From<webp::WebPImage> for ImageMetaResponse {
+    fn from(img: webp::WebPImage) -> Self {
+        Self { width: img.width(), height: img.height(), format: Some("webp") }
+    }
+}
+
+/// Read image dimensions (cheaply), used in `get_image_metadata()`, supports SVG
+pub fn read_image_metadata<P: AsRef<Path>>(path: P) -> Result<ImageMetaResponse> {
+    let path = path.as_ref();
+    let ext = path.extension().and_then(OsStr::to_str).unwrap_or("").to_lowercase();
+
+    let error = |e: Box<dyn StdError + Send + Sync>| {
+        Error::chain(format!("Failed to read image: {}", path.display()), e)
+    };
+
+    match ext.as_str() {
+        "svg" => {
+            let img = SvgMetadata::parse_file(&path).map_err(|e| error(e.into()))?;
+            match (img.height(), img.width(), img.view_box()) {
+                (Some(h), Some(w), _) => Ok((h, w)),
+                (_, _, Some(view_box)) => Ok((view_box.height, view_box.width)),
+                _ => Err("Invalid dimensions: SVG width/height and viewbox not set.".into()),
+            }
+            .map(|(h, w)| ImageMetaResponse::new_svg(h as u32, w as u32))
+        }
+        "webp" => {
+            // Unfortunatelly we have to load the entire image here, unlike with the others :|
+            let data = fs::read(path).map_err(|e| error(e.into()))?;
+            let decoder = webp::Decoder::new(&data[..]);
+            decoder
+                .decode()
+                .map(ImageMetaResponse::from)
+                .ok_or_else(|| Error::msg(format!("Failed to decode WebP image: {}", path.display())))
+        }
+        _ => ImageMeta::read(path).map(ImageMetaResponse::from).map_err(|e| error(e.into())),
+    }
+}
+
+/// Assert that `address` matches `prefix` + RESIZED_FILENAME regex + "." + `extension`,
+/// this is useful in test so that we don't need to hardcode hash, which is annoying.
+pub fn assert_processed_path_matches(path: &str, prefix: &str, extension: &str) {
```

Comment:
can that be moved to tests or with a cfg[test]?

Reply:
`cfg[test]` can't be used here as integration tests build against a non-test library, iirc. But it could be moved to tests maybe, yeah, I'll try...